src/Service/U2fService.php

```diff
@@ -0,0 +1,221 @@
+
+        /** @var YubicoSignRequest $yubicoRequest */
+        list($yubicoRequest) = $this->u2fService->getAuthenticateData([$yubicoRegistration]);
+
+        $request = new SignRequest();
+        $request->version   = $yubicoRequest->version;
+        $request->challenge = $yubicoRequest->challenge;
+        $request->appId     = $yubicoRequest->appId;
+        $request->keyHandle = $yubicoRequest->keyHandle;
+
+        return $request;
+    }
+
+    /**
+     * Request signing of a sign request. Does not support U2F's sign counter system.
+     *
+     * @param Registration $registration The registration that is to be signed.
+     * @param SignRequest  $request The sign request that you requested earlier and was used to query the U2F device.
+     * @param SignResponse $response The response that the U2F device gave in response to the sign request.
+     * @return AuthenticationVerificationResult
+     */
+    public function verifyAuthentication(Registration $registration, SignRequest $request, SignResponse $response)
+    {
+        if (!$this->appId->equals(new AppId($request->appId))) {
+            return AuthenticationVerificationResult::appIdMismatch();
+        }
+
+        if ($response->errorCode !== 0) {
+            return AuthenticationVerificationResult::deviceReportedError($response->errorCode);
+        }
+
+        $yubicoRegistration = new YubicoRegistration();
+        $yubicoRegistration->keyHandle = $registration->keyHandle;
+        $yubicoRegistration->publicKey = $registration->publicKey;
+        $yubicoRegistration->counter   = $registration->signCounter;
+
+        $yubicoRequest = new YubicoSignRequest();
+        $yubicoRequest->version   = $request->version;
+        $yubicoRequest->challenge = $request->challenge;
+        $yubicoRequest->appId     = $request->appId;
+        $yubicoRequest->keyHandle = $request->keyHandle;
+
+        $yubicoResponse = new \stdClass;
+        $yubicoResponse->keyHandle     = $response->keyHandle;
+        $yubicoResponse->signatureData = $response->signatureData;
+        $yubicoResponse->clientData    = $response->clientData;
+
+        try {
+            $yubicoRegistration = $this->u2fService->doAuthenticate(
+                [$yubicoRequest],
+                [$yubicoRegistration],
+                $yubicoResponse
+            );
+        } catch (Error $error) {
+            switch ($error->getCode()) {
+                case \u2flib_server\ERR_NO_MATCHING_REQUEST:
+                    return AuthenticationVerificationResult::requestResponseMismatch();
+                case \u2flib_server\ERR_NO_MATCHING_REGISTRATION:
+                    return AuthenticationVerificationResult::responseRegistrationMismatch();
+                case \u2flib_server\ERR_PUBKEY_DECODE:
+                    return AuthenticationVerificationResult::publicKeyDecodingFailed();
+                case \u2flib_server\ERR_AUTHENTICATION_FAILURE:
```

Comment:
Seems you're missing a rather important one, the [counter too low][1] :wink: 

[1]: https://github.com/Yubico/php-u2flib-server/blob/ecc7447fc6d9bf75259e0fc2ef9f5357341c61f9/src/u2flib_server/U2F.php#L304